### PR TITLE
[MAINTENANCE] `DataContextVariableKey` for use in Stores

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Update to `DatasourceStore` to utilize a new key called `DataContextVariableKey`
- This key adheres to a particular schema and allows devs to specify both the resource type and the name of that resource (as applicable). It provides more flexibility than the existing `StringKey` that was previously used.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/data_context_key.py [hunk 1]
```diff
@@ -74,3 +79,25 @@ def to_fixed_length_tuple(self):
...
+
+
+class DataContextVariableKey(DataContextKey):
+    def __init__(
+        self,
+        resource_type: DataContextVariableSchema,
+        resource_name: Optional[str] = None,
+    ) -> None:
+        self._resource_type = resource_type
+        self._resource_name = resource_name
```
Comment (cdkini) [inline great_expectations/core/data_context_key.py:91]: Some `DataContextVariables` are strings or other single values while other ones are dictionaries.

If I want to use my `InlineStoreBackend` to add a new datasource or store, I want the ability to specify my type (i.e. datasource or store) as well as the name associated with the new object (i.e. my_new_datasource or my_new_store).

The old `StringKey` did not allow for this flexibility.

File: great_expectations/data_context/types/data_context_variables.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    CHECKPOINT_STORE_NAME = "checkpoint_store_name"
+    PROFILER_STORE_NAME = "profiler_store_name"
+    PLUGINS_DIRECTORY = "plugins_directory"
+    VALIDATION_OPERATORS = "validation_operators"
+    STORES = "stores"
+    DATA_DOCS_SITES = "data_docs_sites"
+    NOTEBOOKS = "notebooks"
+    CONFIG_VARIABLES_FILE_PATH = "config_variables_file_path"
+    ANONYMIZED_USAGE_STATISTICS = "anonymous_usage_statistics"
+    STORE_BACKEND_DEFAULTS = "store_backend_defaults"
```
Comment (anthonyburdi) [inline great_expectations/data_context/types/data_context_variables.py]:
```suggestion
```

File: great_expectations/data_context/types/data_context_variables.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+class DataContextVariableSchema(str, enum.Enum):
+    CONFIG_VERSION = "config_version"
+    DATASOURCES = "datasources"
+    EXPECTATIONS_STORE_NAME = "expectations_store_name"
+    VALIDATIONS_STORE_NAME = "validations_store_name"
+    EVALUATION_PARAMETER_STORE_NAME = "evaluation_parameter_store_name"
+    CHECKPOINT_STORE_NAME = "checkpoint_store_name"
+    PROFILER_STORE_NAME = "profiler_store_name"
+    PLUGINS_DIRECTORY = "plugins_directory"
+    VALIDATION_OPERATORS = "validation_operators"
```
Comment (anthonyburdi) [inline great_expectations/data_context/types/data_context_variables.py]:
```suggestion
```

File: scripts/check_docstring_coverage.py [hunk 4]
```diff
@@ -7,7 +7,7 @@
 Diagnostics = Dict[str, List[Tuple[ast.FunctionDef, bool]]]
 
 DOCSTRING_ERROR_THRESHOLD: int = (
-    1100  # This number is to be reduced as we document more public functions!
+    1098  # This number is to be reduced as we document more public functions!
```
Comment (anthonyburdi) [inline scripts/check_docstring_coverage.py:10]: ❤️

File: tests/data_context/store/test_store_backends.py [hunk 5]
```diff
@@ -1556,25 +1560,31 @@ def test_InlineStoreBackend(empty_data_context: DataContext) -> None:
...
 
     assert empty_data_context.config.config_version == new_config_version
     assert mock_save.call_count == 1
 
     # test .get
-    ret = inline_store_backend.get(("config_version",))
+    key = DataContextVariableKey(resource_type=DataContextVariableSchema.CONFIG_VERSION)
+    tuple_ = key.to_tuple()
+    ret = inline_store_backend.get(tuple_)
     assert ret == new_config_version
```
Comment (anthonyburdi) [inline tests/data_context/store/test_store_backends.py:1591]: This test is great - but I do think we also need an integration test to just make sure we read/write to disk (using a mocked filesystem e.g. `pyfakefs`, can be in a separate PR).

File: great_expectations/data_context/store/inline_store_backend.py [hunk 6]
```diff
@@ -1,6 +1,10 @@
 from typing import Any, List, Optional, Tuple
 
 from great_expectations.data_context.store.store_backend import StoreBackend
+from great_expectations.data_context.types.base import DataContextConfig
+from great_expectations.data_context.types.data_context_variables import (
+    DataContextVariableSchema,
+)
 from great_expectations.exceptions.exceptions import StoreBackendError
 from great_expectations.util import filter_properties_dict
```
Comment (anthonyburdi) [inline great_expectations/data_context/store/inline_store_backend.py:10]: I know this code wasn't modified in this PR, and can be done in a different PR but we should have a docstring for the class that describes what an InlineStoreBackend is. Including mentioning that it is only for use with file-backed data contexts (`DataContext` and `FileDataContext`).

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Thank you for this PR 😄  Just a few small items to address.

File: great_expectations/data_context/types/data_context_variables.py (hunk 3 above)
Comment (cdkini) [reply great_expectations/data_context/types/data_context_variables.py]: done!

File: great_expectations/data_context/types/data_context_variables.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/data_context/types/data_context_variables.py]: done!

File: scripts/check_docstring_coverage.py (hunk 4 above)
Comment (cdkini) [reply scripts/check_docstring_coverage.py:10]: slowly getting there!

File: great_expectations/data_context/store/inline_store_backend.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/data_context/store/inline_store_backend.py:10]: I definitely agree - should have added this during my initial implementation!

Updated but please let me know if you think anything else should be added.

File: tests/data_context/store/test_store_backends.py (hunk 5 above)
Comment (cdkini) [reply tests/data_context/store/test_store_backends.py:1591]: done!

**Review by anthonyburdi — APPROVED** [review]
LGTM!
